pkg/tbtc/inactivity_submit.go

```diff
@@ -0,0 +1,209 @@
+	if claim == nil {
+		return nil, fmt.Errorf("claim is nil")
+	}
+
+	claimHash, err := ics.chain.CalculateInactivityClaimSignatureHash(claim)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"inactivity claim hash calculation failed [%w]",
+			err,
+		)
+	}
+
+	signing := ics.chain.Signing()
+
+	signature, err := signing.Sign(claimHash[:])
+	if err != nil {
+		return nil, fmt.Errorf(
+			"inactivity claim hash signing failed [%w]",
+			err,
+		)
+	}
+
+	return &inactivity.SignedClaim{
+		PublicKey: signing.PublicKey(),
+		Signature: signature,
+		ClaimHash: claimHash,
+	}, nil
+}
+
+// VerifySignature verifies if the signature was generated from the provided
+// inactivity claim using the provided public key.
+func (ics *inactivityClaimSigner) VerifySignature(
+	signedClaim *inactivity.SignedClaim,
+) (
+	bool,
+	error,
+) {
+	return ics.chain.Signing().VerifyWithPublicKey(
+		signedClaim.ClaimHash[:],
+		signedClaim.Signature,
+		signedClaim.PublicKey,
+	)
+}
+
+type inactivityClaimSubmitter struct {
+	inactivityLogger log.StandardLogger
+
+	chain           Chain
+	groupParameters *GroupParameters
+	groupMembers    []uint32
+
+	waitForBlockFn waitForBlockFn
+}
+
+func newInactivityClaimSubmitter(
+	inactivityLogger log.StandardLogger,
+	chain Chain,
+	groupParameters *GroupParameters,
+	groupMembers []uint32,
+	waitForBlockFn waitForBlockFn,
+) *inactivityClaimSubmitter {
+	return &inactivityClaimSubmitter{
+		inactivityLogger: inactivityLogger,
+		chain:            chain,
+		groupParameters:  groupParameters,
+		groupMembers:     groupMembers,
+		waitForBlockFn:   waitForBlockFn,
+	}
+}
+
+func (ics *inactivityClaimSubmitter) SubmitClaim(
+	ctx context.Context,
+	memberIndex group.MemberIndex,
+	claim *inactivity.Claim,
+	signatures map[group.MemberIndex][]byte,
+) error {
+	if len(signatures) < ics.groupParameters.HonestThreshold {
+		return fmt.Errorf(
+			"could not submit inactivity claim with [%v] signatures for "+
+				"group honest threshold [%v]",
+			len(signatures),
+			ics.groupParameters.HonestThreshold,
+		)
+	}
+
+	// The inactivity nonce at the beginning of the execution process.
+	inactivityNonce := claim.Nonce
+
+	walletPublicKeyHash := bitcoin.PublicKeyHash(claim.WalletPublicKey)
+
+	walletRegistryData, err := ics.chain.GetWallet(walletPublicKeyHash)
+	if err != nil {
+		return fmt.Errorf("could not get registry data on wallet: [%v]", err)
+	}
+
+	ecdsaWalletID := walletRegistryData.EcdsaWalletID
+
+	currentNonce, err := ics.chain.GetInactivityClaimNonce(
+		ecdsaWalletID,
+	)
+	if err != nil {
+		return fmt.Errorf("could not get nonce for wallet: [%v]", err)
+	}
+
+	if currentNonce.Cmp(inactivityNonce) > 0 {
+		// Someone who was ahead of us in the queue submitted the claim. Giving up.
+		ics.inactivityLogger.Infof(
+			"[member:%v] inactivity claim already submitted; "+
+				"aborting inactivity claim on-chain submission",
+			memberIndex,
+		)
+		return nil
+	}
+
+	inactivityClaim, err := ics.chain.AssembleInactivityClaim(
```

Comment:
We already have `claim` in this context so let's add the `chain*` prefix to make it clear what we assemble here.
```suggestion
	chainClaim, err := ics.chain.AssembleInactivityClaim(
```

Reply:
Done in 0e3b9e5a5e7bfe2b010ecf77509bdfe5536713a2.